src/request/mod.rs

```diff
@@ -22,12 +24,103 @@ pub mod plan;
 mod plan_builder;
 #[macro_use]
 mod shard;
+pub mod codec;
 
 /// Abstracts any request sent to a TiKV server.
 #[async_trait]
-pub trait KvRequest: Request + Sized + Clone + Sync + Send + 'static {
+pub trait KvRequest<C>: Request + Sized + Clone + Sync + Send + 'static {
     /// The expected response to the request.
-    type Response: HasKeyErrors + HasLocks + Clone + Send + 'static;
+    type Response: HasKeyErrors + HasLocks + HasRegionError + Clone + Send + 'static;
+
+    fn encode_request(self, _codec: &C) -> Self {
+        self
+    }
+
+    fn decode_response(&self, _codec: &C, resp: Self::Response) -> crate::Result<Self::Response> {
+        Ok(resp)
+    }
+}
+
+#[macro_export]
+macro_rules! impl_kv_request_for_single_key_op {
+    ($req: ty, $resp: ty) => {
+        impl<C> KvRequest<C> for $req
+        where
+            C: RequestCodec,
+        {
+            type Response = $resp;
+
+            fn encode_request(mut self, codec: &C) -> Self {
+                *self.mut_key() = codec.encode_key(self.take_key());
+
+                self
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! impl_kv_request_for_batch_get {
+    ($req: ty, $resp: ty) => {
+        impl<C> KvRequest<C> for $req
+        where
+            C: RequestCodec,
+        {
+            type Response = $resp;
+
+            fn encode_request(mut self, codec: &C) -> Self {
+                *self.mut_keys() = codec.encode_keys(self.take_keys());
+
+                self
+            }
+
+            fn decode_response(
+                &self,
+                codec: &C,
+                mut resp: Self::Response,
+            ) -> $crate::Result<Self::Response> {
+                *resp.mut_pairs() = codec.decode_pairs(resp.take_pairs())?;
+
+                Ok(resp)
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! impl_kv_request_for_scan_op {
```

Comment:
These macro rules have the same patterns which is to take some field out and then encode or decode it. We should extract them into one rule.